packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -459,21 +459,54 @@ export class Pipeline extends PipelineBase {
 
     // check whether we already have a bucket in that region,
     // either passed from the outside or previously created
+    const crossRegionSupport = this.obtainCrossRegionSupportFor(action, actionEnv.region);
+
+    // the stack containing the replication bucket must be deployed before the pipeline
+    Stack.of(this).addDependency(crossRegionSupport.stack);
+    // The Pipeline role must be able to replicate to that bucket
+    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+
+    return {
+      artifactBucket: crossRegionSupport.replicationBucket,
+      region: actionEnv.region,
+    };
+  }
+
+  /**
+   * Get or create the cross-region support construct for the given action
+   */
+  private obtainCrossRegionSupportFor(action: IAction, actionRegion: string) {
     let crossRegionSupport = this._crossRegionSupport[actionRegion];
     if (!crossRegionSupport) {
       // we need to create scaffolding resources for this region
+      const otherStack = this.scopeForCrossRegionSupportConstruct(action);
       crossRegionSupport = this.createSupportResourcesForRegion(otherStack, actionRegion);
       this._crossRegionSupport[actionRegion] = crossRegionSupport;
     }
+    return crossRegionSupport;
+  }
 
-    // the stack containing the replication bucket must be deployed before the pipeline
-    pipelineStack.addDependency(crossRegionSupport.stack);
-    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+  /**
+   * Return the correct scope to create the cross-region construct in
+   *
+   * As an optimization, we reuse the action resource's stack if the stack has
+   * the same environment as the action.
+   *
+   * (Always true for owned action resources but not necessarily true
+   * for imported ones).
+   */
+  private scopeForCrossRegionSupportConstruct(action: IAction): Stack | undefined {
+    // The resource this action is targeting (if any). Might be owned, might be imported.
+    const actionResource = action.actionProperties.resource;
+    if (!actionResource) { return undefined; }
 
-    return {
-      artifactBucket: crossRegionSupport.replicationBucket,
-      region: actionRegion,
-    };
+    const actionResourceStack = Stack.of(actionResource);
+
+    const stackHasCorrectEnvironment =
+      sameEnvDimension(actionResource.env.region, actionResourceStack.region)
+      && sameEnvDimension(actionResource.env.account, actionResourceStack.account);
+
+    return stackHasCorrectEnvironment ? actionResourceStack : undefined;
```

Comment:
Since you hate my naming, let's tackle it a different way.

In my opinion, there's a pretty useful concept here (that of a Resource belonging to the same environment as its owning Stack). So, let's extract a helper function for checking that, and change this to:

```suggestion
    return resourceInSameEnvAsItsStack(actionResource) ? Stack.of(actionResource) : undefined;
```

(If the name still sucks in your opinion, feel free to change it to a better one)

Reply:
I went even more reusable and made a `sameEnv(env1, env2)` function. I think this is neater even.